### DISCUSSION TITLE
How to decorate/wrap atom setter function?

### DISCUSSION THREAD

**Discussion by jmmaa** [pmndrs/jotai#1494] [Q&A] [answered]
Hi, I've been trying the lib for a while and got into a mind block on how to decorate the setter function of an atom.
To make this clearer, here is my approach

```ts
export const throttlerAtom = atom(
  null,
  _.throttle(
    (get: any, set: any, callback: (page: number) => number) => {
      set(pageAtom, callback(get(pageAtom)));
    },
    300,
    { leading: true, trailing: false }
  )
);
```
basically I'm throttling `pageAtom` setter by using  `throttlerAtom` setter. However I also want to access other atoms in outside the scope of `throttle` which will allow me to configure the throttling behavior. Lets say I want to set the wait time for it by having another atom `waitAtom`:
```ts
const waitAtom = atom(2000)
```
and then I want to use it in my throttler atom like this
```ts
export const throttlerAtom = atom(
  null,
  (get, set, callback: (page: number) => number) => {

    const throttled = _.throttle(
      (callback: (page: number) => number) =>
        set(pageAtom, callback(get(pageAtom))),
      get(waitAtom) // atom value for wait time
    );

    throttled(callback);
  }
);
```
The result is that the throttle does not work, but it still was able to set values for `pageAtom`. Are there any workaround for this or any recommended approaches?

**Comment by dai-shi** [comment, marked as answer]
I don't think wrapping a `write` function is an expected/supported usage. (hm, but it does work. interesting.)

The second trial seems good, but you need to define `throttled` outside of the write function.
And, pass those values through the throttled function parameters.
Ah, and you want `waitAtom`. That's tricky.

Something like this:
```ts
const waitAtom = atom(2000)

const throttledAtom = atom((get) => {
  const wait = get(waitAtom)
  const fn = _.throttle((callback: () => void) => callback(), wait);
  return { fn } // wrapping a function may not be necessary, but as a convention.
})

export const throttlerAtom = atom(
  null,
  (get, set, callback: (page: number) => number) => {
    const throttled = get(throttledAtom)
    throttled.fn(() => {
      set(pageAtom, callback(get(pageAtom)))
    })
  }
);
```

**Reply by jmmaa** [reply]
I actually did solve this by decorating the throttled atom dynamically, getting its setter then use it in my other atom setters that needs it, something like this:
```ts
export const useNav = () => {
  const [, nextCallback] = useAtom(goNextAtom);
  const [, prevCallback] = useAtom(goPrevAtom);

  const [wait] = useAtom(waitAtom);

  const [, setIndex] = useAtom(
    React.useMemo(() => throttleSetIndexAtom({ delay: wait }), [])
  );

  const next = React.useCallback(() => nextCallback(setIndex), []);
  const prev = React.useCallback(() => prevCallback(setIndex), []);

  return [next, prev];
};
```
Here is the throttled atom
```ts
export const throttleSetIndexAtom = (options?: {
  delay?: number | undefined;
  options?: _.ThrottleSettings | undefined;
}) =>
  atom(
    null,
    _.throttle(
      (get, set, callback: (index: number) => number) =>
        set(indexAtom, callback),
      options ? options.delay : 0,
      options ? options.options : undefined
    )
  );
```
Thanks for the example, its so much cleaner!
